datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -351,56 +468,108 @@ impl CommonSubexprEliminate {
         let saved_names = aggr_expr
             .iter()
             .map(|expr| name_perserver.save(expr))
             .collect::<Result<Vec<_>>>()?;
 
-        // rewrite both group exprs and aggr_expr
-        let rewritten = self.rewrite_expr(
-            vec![group_expr, aggr_expr],
-            &[&group_arrays, &aggr_arrays],
-            unwrap_arc(input),
-            &expr_stats,
-            config,
-        )?;
-        transformed |= rewritten.transformed;
-        let (mut new_expr, new_input) = rewritten.data;
-
-        // note the reversed pop order.
-        let new_aggr_expr = pop_expr(&mut new_expr)?;
-        let new_group_expr = pop_expr(&mut new_expr)?;
+        let mut expr_stats = ExprStats::new();
+        // rewrite inputs
+        let (group_found_common, group_arrays) =
+            self.to_arrays(&group_expr, &mut expr_stats, ExprMask::Normal)?;
+        let (aggr_found_common, aggr_arrays) =
+            self.to_arrays(&aggr_expr, &mut expr_stats, ExprMask::Normal)?;
+        let (new_aggr_expr, new_group_expr, new_input) =
+            if group_found_common || aggr_found_common {
+                // rewrite both group exprs and aggr_expr
+                let rewritten = self.rewrite_expr(
+                    vec![group_expr.clone(), aggr_expr.clone()],
+                    vec![group_arrays, aggr_arrays],
+                    unwrap_arc(input),
+                    &expr_stats,
+                    config,
+                )?;
+                assert!(rewritten.transformed);
+                transformed |= rewritten.transformed;
+                let (mut new_expr, new_input) = rewritten.data;
+
+                // note the reversed pop order.
+                let new_aggr_expr = pop_expr(&mut new_expr)?;
+                let new_group_expr = pop_expr(&mut new_expr)?;
+
+                (new_aggr_expr, new_group_expr, Arc::new(new_input))
+            } else {
+                (aggr_expr, group_expr, input)
+            };
 
         // create potential projection on top
         let mut expr_stats = ExprStats::new();
-        let new_input_schema = Arc::clone(new_input.schema());
-        let aggr_arrays = to_arrays(
+        let (aggr_found_common, aggr_arrays) = self.to_arrays(
             &new_aggr_expr,
             &mut expr_stats,
             ExprMask::NormalAndAggregates,
         )?;
-        let mut common_exprs = IndexMap::new();
-        let mut rewritten_exprs = self.rewrite_exprs_list(
-            vec![new_aggr_expr.clone()],
-            &[&aggr_arrays],
-            &expr_stats,
-            &mut common_exprs,
-            &config.alias_generator(),
-        )?;
-        transformed |= rewritten_exprs.transformed;
-        let rewritten = pop_expr(&mut rewritten_exprs.data)?;
+        if aggr_found_common {
+            let mut common_exprs = CommonExprs::new();
+            let mut rewritten_exprs = self.rewrite_exprs_list(
+                vec![new_aggr_expr.clone()],
```

Comment:
Extra clone, but the reason is the same as https://github.com/apache/datafusion/pull/10473/files#r1648985272

Reply:
maybe we can add a comment

Reply:
Added in https://github.com/apache/datafusion/pull/10473/commits/ee61224f1a948c38ac99ae53d9d426bb69da9548.